### PR TITLE
Add fast power

### PR DESCRIPTION
**Pull request by ibahadiraltun**
#### Description of Change
Implemented algorithm that computes a^b in O(logN) time recursively or linear. 

#### Checklist
- [x] Added description of change
- [x] Added tests and example, test must pass
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by ibahadiraltun** [comment]
Fast Power algorithm that calculates a^b in O(logN) time.

**Comment by ibahadiraltun** [comment]
Added sample tests

File: math/fast_power.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    return result;
+}
+
+int main() {
+    std::ios_base::sync_with_stdio(false);
+
+    std::cout << "3^5 is = " << fast_power_recursive(3, 5) << std::endl;
+    std::cout << "6^4 is = " << fast_power_recursive(6, 4) << std::endl;
+    std::cout << "2^7 is = " << fast_power_recursive(2, 7) << std::endl;
+    std::cout << "10^3 is = " << fast_power_recursive(10, 3) << std::endl;
```
Comment (cclauss) [inline math/fast_power.cpp]: Would it be possible to create some tests of the form:
```cpp
assert (fast_power_recursive(10, 3) == 1000);
```
that would raise an error if __fast_power_recursive()__ was modified incorrectly?

Should we test zero and negative values in our tests?

Comment (ibahadiraltun) [reply math/fast_power.cpp]: It would be possible to create tests using **assert()**.
This algorithm works for negative base but not for negative power. 
We can use the fact that aˆ(-b) = 1/(aˆb) but I do not think it is often used.

Comment (cclauss) [reply math/fast_power.cpp]: Perhaps a Python example would help...
```

>>> 10 ** 3
1000
>>> -10 ** 3
-1000
>>> 10 ** -3
0.001
>>> -10 ** -3
-0.001
```
When run with each of these values, does the proposed algorithm return similar results?  If not, how does it fail?  Does if produce an invalid result or does it raise a helpful error?

Comment (cclauss) [reply math/fast_power.cpp]: Perhaps tests of the form:
```cpp
assert (fast_power_recursive(10, 3) ==  std::pow(10, 3))
```

Comment (ibahadiraltun) [reply math/fast_power.cpp]: Updated test form.

**Comment by ibahadiraltun** [comment]
Added more random tests in appropriate form and provide a return value for negative powers.

**Comment by cclauss** [comment]
Awesome contribution!  Thanks.

**Comment by ibahadiraltun** [comment]
> Awesome contribution! Thanks.

Thank you!
